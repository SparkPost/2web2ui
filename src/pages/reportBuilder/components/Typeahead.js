import React, { useCallback, useState, useMemo } from 'react';
import { ComboBoxTypeahead } from 'src/components/typeahead/AsyncComboBoxTypeahead';
import { METRICS_API_LIMIT } from 'src/constants';
import sortMatch from 'src/helpers/sortMatch';
import { useSparkPostQuery } from 'src/hooks';

function Typeahead(props) {
  const {
    id,
    setFilterValues,
    index,
    lookaheadRequest,
    lookaheadOptions = {},
    selector,
    value,
    groupingIndex,
    filterIndex,
    filterType,
    ...rest
  } = props;
  const [inputValue, setInputValue] = useState('');

  const { data, status } = useSparkPostQuery(
    () =>
      lookaheadRequest({
        ...lookaheadOptions,
        match: inputValue,
        limit: METRICS_API_LIMIT,
        staleTime: 1000 * 30, // 30 seconds
      }),
    { enabled: Boolean(inputValue && inputValue.length >= 3), refetchOnWindowFocus: false },
  );

  const results = selector(data);

  const onFilterChange = useCallback(
    newValues => {
      setFilterValues({ groupingIndex, filterIndex, values: newValues });
    },
    [groupingIndex, filterIndex, setFilterValues],
  );

  const filteredResults = useMemo(() => {
    return sortMatch(results, inputValue, a => a.value);
  }, [inputValue, results]);

  return (
    <ComboBoxTypeahead
      id={id}
      onChange={onFilterChange}
      onInputChange={setInputValue}
      value={value}
      results={filteredResults}
      loading={status === 'loading'}
      customItemHelpText="Add Filter"
      {...rest}
    />
  );
}

export default Typeahead;
