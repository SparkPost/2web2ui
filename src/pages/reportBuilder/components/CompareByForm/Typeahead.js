import React, { useEffect, useState, useMemo } from 'react';
import { useDebouncedCallback } from 'use-debounce';
import Downshift from 'downshift';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { KeyboardArrowDown, KeyboardArrowUp } from '@sparkpost/matchbox-icons';
import { METRICS_API_LIMIT } from 'src/constants';
import sortMatch from 'src/helpers/sortMatch';
import { ActionList, Box, Inline, TextField } from 'src/components/matchbox';
import { LoadingSVG } from 'src/components';
import { useSparkPostQuery } from 'src/hooks';

const Loading = () => <LoadingSVG size="XSmall" />;

const PopoverActionList = styled(ActionList)`
  background: ${props => props.theme.colors.white};
  border: 1px solid ${props => props.theme.colors.gray['400']};
  box-shadow: ${props => props.theme.shadows['200']};
  left: 0;
  margin-top: 4px;
  max-height: 20rem;
  opacity: ${props => (props.open ? 1 : 0)};
  overflow-y: scroll;
  pointer-events: ${props => (props.open ? 'auto' : 'none')};
  visibility: ${props => (props.open ? 'visible' : 'hidden')};
  position: absolute;
  right: 0;
  top: 100%;
  transition: opacity ${props => props.theme.motion.duration['fast']} ease-out;
  z-index: ${props => props.theme.zIndices['overlay'] + 1};
`;

const TypeaheadWrapper = styled.div`
  margin: 0 0 1rem;
  position: relative;
  &:last-child {
    margin-bottom: 0;
  }
`;

function TypeSelect({
  disabled,
  helpText,
  id,
  itemToString,
  label,
  maxHeight,
  maxWidth,
  onChange,
  placeholder,
  results,
  selectedItem,
  onInputChange,
  loading,
  suffix,
  labelHidden,
  customOptionHelpText,
}) {
  const [matches, setMatches] = useState([]);
  // Controlled input so that we can change the value after selecting dropdown.
  const [inputValue, setInputValue] = useState(selectedItem ? itemToString(selectedItem) : '');
  const [debouncedValue, setDebouncedValue] = useState('');
  const [debounceInputChange] = useDebouncedCallback(setDebouncedValue, 300);

  useEffect(() => {
    onInputChange(debouncedValue);
  }, [onInputChange, debouncedValue]);

  useEffect(() => {
    setMatches(results);
  }, [results]);

  useEffect(() => {
    debounceInputChange(inputValue);
  }, [debounceInputChange, inputValue]);

  const handleStateChange = (changes, downshift) => {
    // Highlights first item in list by default
    if (!downshift.highlightedIndex) {
      downshift.setHighlightedIndex(0);
    }
  };

  useEffect(() => {
    if (selectedItem) {
      const selectedValue = itemToString(selectedItem);
      if (selectedValue !== inputValue) {
        setInputValue(selectedValue);
      }
    } else if (selectedItem === null) {
      setInputValue('');
    }
    // eslint-disable-next-line
  }, [selectedItem]);

  const typeaheadFn = ({
    getInputProps,
    getItemProps,
    getMenuProps,
    highlightedIndex,
    inputValue,
    isOpen,
    selectedItem,
  }) => {
    const textFieldConfig = {
      disabled,
      label,
      labelHidden,
      placeholder,
      helpText,
      onBlur: () => {
        setInputValue(itemToString(selectedItem));
      },
      onFocus: event => {
        event.target.select();
      },
    };

    const isMenuOpen = isOpen && Boolean(debouncedValue);
    const textFieldProps = getInputProps(textFieldConfig);
    textFieldProps['data-lpignore'] = true;

    const SuffixIcon = isOpen ? KeyboardArrowUp : KeyboardArrowDown;

    // note, this is identifical behavior as AsyncComboBoxTypeahead
    const customItem = {
      id: inputValue.trim(),
      type: label,
      value: inputValue.trim(),
    };

    const looksLikeCustomItem = item => {
      // hack, try our best to match input to items
      if (item.id) {
        return (
          String(item.id) === customItem.id ||
          item.value === customItem.value ||
          item.value === `${customItem.value} (ID ${item.id})`
        );
      }

      return item.value === customItem.value;
    };

    const items = [
      {
        content: `"${customItem.value}"`,
        helpText: customOptionHelpText,
        isVisible: customItem.value.length > 0 && !matches.some(looksLikeCustomItem),
        item: customItem,
      },
      ...matches.map(item => ({
        content: itemToString(item),
        isVisible: true,
        item,
      })),
    ]
      .filter(({ isVisible }) => isVisible)
      .map((item, index) =>
        getItemProps({
          ...item,
          highlighted: highlightedIndex === index,
          index,
          key: `${id}_item_${index}`,
        }),
      );

    return (
      <div>
        <TypeaheadWrapper>
          <Box maxWidth={maxWidth} position="relative">
            {/* hack, can't apply getMenuProps to ActionList because ref prop is not supported */}
            <div {...getMenuProps()}>
              <PopoverActionList open={isMenuOpen} maxHeight={maxHeight}>
                {Boolean(items.length) ? (
                  items.map(item => (
                    <ActionList.Action {...item}>
                      <TypeaheadItem label={item.content} />
                    </ActionList.Action>
                  ))
                ) : (
                  <Box p="200">No Results</Box>
                )}

                {loading && (
                  <Box ml="200">
                    <Loading />
                  </Box>
                )}
              </PopoverActionList>
            </div>
            <TextField
              suffix={suffix !== undefined ? suffix : <SuffixIcon color="blue.700" size={25} />}
              {...textFieldProps}
            />
          </Box>
        </TypeaheadWrapper>
      </div>
    );
  };

  return (
    <Downshift
      id={id}
      inputValue={inputValue}
      itemToString={itemToString}
      onChange={onChange}
      onInputValueChange={setInputValue}
      onStateChange={handleStateChange}
      selectedItem={selectedItem}
    >
      {typeaheadFn}
    </Downshift>
  );
}

TypeSelect.propTypes = {
  customOptionHelpText: PropTypes.string,
  disabled: PropTypes.bool,
  id: PropTypes.string.isRequired,
  itemToString: PropTypes.func,
  label: PropTypes.string,
  maxHeight: PropTypes.number,
  maxNumberOfResults: PropTypes.number,
  maxWidth: PropTypes.number,
  onChange: PropTypes.func,
  placeholder: PropTypes.string,
  results: PropTypes.array,
};

TypeSelect.defaultProps = {
  disabled: false,
  itemToString: item => item,
  maxHeight: 300,
  maxNumberOfResults: 100,
  maxWidth: 1200,
  placeholder: 'Type to search',
  results: [],
};

function TypeaheadItem({ label }) {
  return <Inline as="span">{label}</Inline>;
}

TypeaheadItem.propTypes = {
  label: PropTypes.string.isRequired,
  meta: PropTypes.string,
};

function Typeahead({ id, onChange, lookaheadRequest, lookaheadOptions, selector, ...rest }) {
  const [inputValue, setInputValue] = useState('');

  const { data, status } = useSparkPostQuery(
    () =>
      lookaheadRequest({
        ...lookaheadOptions,
        match: inputValue,
        limit: METRICS_API_LIMIT,
      }),
    {
      enabled: Boolean(inputValue && inputValue.length >= 3),
      refetchOnWindowFocus: false,
      staleTime: 1000 * 30, // 30 seconds
    },
  );
  const results = selector(data);
  const filteredResults = useMemo(() => {
    return sortMatch(results, inputValue, a => a.value);
  }, [inputValue, results]);

  return (
    <TypeSelect
      id={id}
      onChange={onChange}
      onInputChange={setInputValue}
      itemToString={item => {
        return item ? item.value : '';
      }}
      loading={status === 'loading'}
      suffix={null}
      {...rest}
      results={filteredResults}
    />
  );
}

export default Typeahead;
