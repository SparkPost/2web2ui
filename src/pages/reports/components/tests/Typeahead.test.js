import React from 'react';
import { shallow } from 'enzyme';
import { Typeahead } from '../Typeahead';

jest.mock('lodash/debounce', () =>
  jest.fn(fn => {
    fn.cancel = jest.fn();
    return fn;
  }),
);

describe('Component: Typeahead', () => {
  let props;
  let wrapper;
  let downshiftHelpers;
  const defaultCrossMatches = [
    { type: 'Sending Domain', value: 'cross' },
    { type: 'Template', value: 'crossing' },
  ];

  beforeEach(() => {
    props = {
      items: [
        { type: 'Sending Domain', value: 'cross' },
        { type: 'Campaign', value: 'treasure' },
        { type: 'Recipient Domain', value: 'cross.example.com' },
        { type: 'Template', value: 'crossing' },
      ],
      matches: [],
      onSelect: jest.fn(),
      refreshTypeaheadCache: jest.fn(() => Promise.resolve()),
      placeholder: '',
      pattern: null,
    };

    downshiftHelpers = {
      getInputProps: jest.fn(),
      getItemProps: jest.fn(),
      getLabelProps: jest.fn(),
      isOpen: true,
      selectedItem: null,
      highlightedIndex: 0,
      clearSelection: jest.fn(),
    };

    wrapper = shallow(<Typeahead {...props} />);
  });

  function renderTypeahead(wrapper, input) {
    const TypeaheadRender = wrapper.instance().onTypeahead;
    return shallow(<TypeaheadRender {...downshiftHelpers} inputValue={input} {...props} />);
  }

  it('should render ok by default', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should trigger on field change', () => {
    wrapper.simulate('inputValueChange', 'cros');
    expect(props.refreshTypeaheadCache).toHaveBeenCalledTimes(1);
  });

  it('should produce matches on search', () => {
    wrapper.simulate('inputValueChange', 'cross');
    expect(wrapper.state().matches).toEqual(defaultCrossMatches);
  });

  it('should produce no matches on empty query', () => {
    wrapper.simulate('inputValueChange', '');
    expect(wrapper.state().matches).toEqual([]);
  });

  it('should first show synchronous matches while waiting for async calls to finish', () => {
    const input = 'cross';
    wrapper.simulate('inputValueChange', input);
    expect(wrapper.state().calculatingMatches).toEqual(true);
    expect(renderTypeahead(wrapper, input)).toMatchSnapshot();
  });

  it('should not make any api requests nor have any matches with a pattern <2 characters', () => {
    const input = 'c';
    return wrapper
      .instance()
      .updateLookAhead(input)
      .then(() => {
        expect(props.refreshTypeaheadCache).not.toHaveBeenCalled();
        expect(renderTypeahead(wrapper, input)).toMatchSnapshot();
      });
  });

  it('should show sync and async matches', () => {
    const input = 'cross';
    return wrapper
      .instance()
      .updateLookAhead(input)
      .then(() => {
        expect(renderTypeahead(wrapper, input)).toMatchSnapshot();
      });
  });

  it('should not update matches on receiving out-of-date results', () => {
    const input = 'cros';
    const promise = wrapper.instance().updateLookAhead(input);
    wrapper.instance().setState({ pattern: 'something else' });
    return promise.then(() => {
      expect(wrapper.state().matches).toEqual(defaultCrossMatches);
    });
  });

  it('should call onSelect on change', () => {
    const value = { type: 'Template', value: 'treasure' };
    wrapper.instance().handleDownshiftChange(value);
    expect(props.onSelect).toHaveBeenCalledWith(value);
  });

  it('should not call onSelect on clear', () => {
    wrapper.instance().handleDownshiftChange(null);
    expect(props.onSelect).not.toHaveBeenCalled();
  });

  it('should cancel any debounced lookahead updates', () => {
    const cancel = jest.spyOn(wrapper.instance().updateLookAhead, 'cancel');
    wrapper.unmount();
    expect(cancel).toHaveBeenCalled();
  });
});
