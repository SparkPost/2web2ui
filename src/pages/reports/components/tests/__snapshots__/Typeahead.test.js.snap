// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Typeahead should first show synchronous matches while waiting for async calls to finish 1`] = `
<div
  className="Typeahead"
>
  <div
    className="List open"
    data-id="report-filters-dropdown"
  >
    <ActionList
      actions={
        Array [
          Object {
            "className": "",
            "content": <TypeaheadItem
              helpText="Sending Domain"
              value="cross"
            />,
            "highlighted": true,
          },
          Object {
            "className": "",
            "content": <TypeaheadItem
              helpText="Template"
              value="crossing"
            />,
            "highlighted": false,
          },
        ]
      }
      maxHeight={300}
    />
  </div>
  <label
    className="Label"
  >
    Reports Filter Typeahead
  </label>
  <TextField
    id="report-options-typeahead"
    suffix={
      <MatchesLoading
        isLoading={true}
      />
    }
  />
</div>
`;

exports[`Component: Typeahead should not make any api requests nor have any matches with a pattern <2 characters 1`] = `
<div
  className="Typeahead"
>
  <div
    className="List"
    data-id="report-filters-dropdown"
  >
    <ActionList
      actions={Array []}
      maxHeight={300}
    />
  </div>
  <label
    className="Label"
  >
    Reports Filter Typeahead
  </label>
  <TextField
    id="report-options-typeahead"
    suffix={
      <MatchesLoading
        isLoading={false}
      />
    }
  />
</div>
`;

exports[`Component: Typeahead should render ok by default 1`] = `
<Downshift
  defaultHighlightedIndex={null}
  defaultIsOpen={false}
  environment={[Window]}
  getA11yStatusMessage={[Function]}
  itemToString={[Function]}
  onChange={[Function]}
  onInputValueChange={[Function]}
  onOuterClick={[Function]}
  onSelect={[Function]}
  onStateChange={[Function]}
  onUserAction={[Function]}
  scrollIntoView={[Function]}
  selectedItemChanged={[Function]}
  stateReducer={[Function]}
  suppressRefError={false}
>
  <Component />
</Downshift>
`;

exports[`Component: Typeahead should show sync and async matches 1`] = `
<div
  className="Typeahead"
>
  <div
    className="List open"
    data-id="report-filters-dropdown"
  >
    <ActionList
      actions={
        Array [
          Object {
            "className": "",
            "content": <TypeaheadItem
              helpText="Sending Domain"
              value="cross"
            />,
            "highlighted": true,
          },
          Object {
            "className": "",
            "content": <TypeaheadItem
              helpText="Recipient Domain"
              value="cross.example.com"
            />,
            "highlighted": false,
          },
          Object {
            "className": "",
            "content": <TypeaheadItem
              helpText="Template"
              value="crossing"
            />,
            "highlighted": false,
          },
        ]
      }
      maxHeight={300}
    />
  </div>
  <label
    className="Label"
  >
    Reports Filter Typeahead
  </label>
  <TextField
    id="report-options-typeahead"
    suffix={
      <MatchesLoading
        isLoading={false}
      />
    }
  />
</div>
`;
